File: send_trajectory.py
from IPython import embed
from geometry_msgs.msg import PoseArray
from geometry_msgs.msg import Pose
from tf_conversions import posemath
import PyKDL
import rospy
from dvrk import psm
import numpy as np
import ipdb

posePub = rospy.Publisher(name = 'set_continuous_palpation_trajectory',
                          data_class = PoseArray, 
                          queue_size = 1)

def send_trajectory(positions, robot):
    ''' 
    SEND_TRAJECTORY Sends a trajectory message to a continuous palpation node
        Positions is a list of PyKDL vectors. The current orientation is kept
        constant.
    '''
    msg = PoseArray()
    msg.header.stamp = rospy.Time.now()
    for idx in range(0,len(positions)):
        pose = posemath.toMsg(robot.get_current_position())
        pose.position.x = positions[idx].x()
        pose.position.y = positions[idx].y()
        pose.position.z = positions[idx].z()
        msg.poses.append(pose)
    posePub.publish(msg)

def send_trajectory_relative(positions, robot):
    ''' 
    SEND_TRAJECTORY Sends a trajectory message to a continuous palpation node
        Positions is a list of PyKDL vectors. The current orientation is kept
        constant.
    '''

    msg = PoseArray()
    msg.header.stamp = rospy.Time.now()
    for idx in range(0,len(positions)):
        pose = posemath.toMsg(robot.get_current_position())
        pose.position.x = pose.position.x + positions[idx].x()
        pose.position.y = pose.position.y + positions[idx].y()
        pose.position.z = pose.position.z + positions[idx].z()
        msg.poses.append(pose)
    posePub.publish(msg)


if __name__ == '__main__':
    robot = psm("PSM1")
    embed()

File: Matlab_scripts/send_trajectory.m
function send_trajectory( positions, robot,publisher)
%SEND_TRAJECTORY Sends a trajectory message to a continuous palpation node
%   Positions is a 3xN matrix of positions. The current orientation is kept
%   constant.
    msg = rosmessage(publisher);
    msg.Header.Stamp = rostime('now');
    curr = robot.position_current_subscriber.LatestMessage.Pose;
    for idx = 1:size(positions,2)
        pose = curr.copy();
        pose.Position.X = positions(1,idx);
        pose.Position.Y = positions(2,idx);
        pose.Position.Z = positions(3,idx);
        msg.Poses = [msg.Poses; pose];
    end
    send(publisher, msg);
end

    